### PR TITLE
catch bad alloc exception

### PR DESCRIPTION
**Pull request by phlrain**
### PR types
Bug fixes

### PR changes
Others

### Describe
修复PE再遇到bad alloc的异常时，异常处理器能够正确捕捉异常

### CODE REVIEW

File: paddle/fluid/framework/details/exception_holder.h [hunk 1]
```diff
@@ -95,6 +106,12 @@ class ExceptionHolder {
     type_ = kEnforceNotMet;
   }
 
+  void Catch(const memory::allocation::BadAlloc& exp) {
+    std::lock_guard<std::mutex> lock(mu_);
+    exception_.reset(new paddle::memory::allocation::BadAlloc(exp));
```
Comment (zhiqiu) [inline paddle/fluid/framework/details/exception_holder.h:123]: Better Add`if (exception_.get() == nullptr)`, do not cover the existing exception.

File: paddle/fluid/framework/details/exception_holder.h [hunk 2]
```diff
@@ -33,6 +34,8 @@ class ExceptionHolder {
       Catch(exp);
     } catch (platform::EnforceNotMet& exp) {
       Catch(exp);
+    } catch (memory::allocation::BadAlloc& exp) {
+      Catch(exp);
     } catch (std::exception& ex) {
```
Comment (zhiqiu) [inline paddle/fluid/framework/details/exception_holder.h:40]: We need to catch all types of exception to Exception_holder

File: paddle/fluid/framework/details/exception_holder.h [hunk 3]
```diff
@@ -33,9 +34,10 @@ class ExceptionHolder {
       Catch(exp);
     } catch (platform::EnforceNotMet& exp) {
       Catch(exp);
+    } catch (memory::allocation::BadAlloc& exp) {
```
Comment (zhiqiu) [inline paddle/fluid/framework/details/exception_holder.h]: Better put `BadAlloc` to the first catch.

File: paddle/fluid/framework/details/exception_holder.h [hunk 4]
```diff
@@ -33,9 +34,10 @@ class ExceptionHolder {
...
+      Catch(exp);
     } catch (std::exception& ex) {
-      PADDLE_THROW(platform::errors::Fatal(
-          "Unknown std::exception caught:\n%s.", ex.what()));
+      Catch(ex);
     } catch (...) {
       PADDLE_THROW(platform::errors::Fatal("Unknown exception caught."));
```
Comment (zhiqiu) [inline paddle/fluid/framework/details/exception_holder.h]: LOG(FETAL) if `...` if not caught by ExceptionHolder

File: paddle/fluid/framework/details/exception_holder.h [hunk 5]
```diff
@@ -95,6 +112,13 @@ class ExceptionHolder {
     type_ = kEnforceNotMet;
   }
 
+  void Catch(const memory::allocation::BadAlloc& exp) {
+    std::lock_guard<std::mutex> lock(mu_);
+    // BadAlloc have the highest priority
+    exception_.reset(new paddle::memory::allocation::BadAlloc(exp));
```
Comment (zhiqiu) [inline paddle/fluid/framework/details/exception_holder.h:123]: Better add log here so we know what `exception_` is before and after reset.

**Review by zhiqiu — DISMISSED** [review]
LGTM

**Review by chenwhql — DISMISSED** [review]

**Review by zhiqiu — APPROVED** [review]

**Review by chenwhql — APPROVED** [review]

**Pull request merged**